plugins/base/src/main/kotlin/transformers/documentables/DocumentableVisibilityFilterTransformer.kt

```diff
@@ -20,10 +21,31 @@ class DocumentableVisibilityFilterTransformer(val context: DokkaContext) : PreMe
     ) {
         fun Visibility.isAllowedInPackage(packageName: String?) = when (this) {
             is JavaVisibility.Public,
-            is KotlinVisibility.Public -> true
-            else -> packageName != null
-                    && packageOptions.firstOrNull { Regex(it.matchingRegex).matches(packageName) }?.includeNonPublic
-                    ?: globalOptions.includeNonPublic
+            is KotlinVisibility.Public -> isAllowedInPackage(packageName, DokkaConfiguration.Visibility.PUBLIC)
+            is JavaVisibility.Private,
+            is KotlinVisibility.Private -> isAllowedInPackage(packageName, DokkaConfiguration.Visibility.PRIVATE)
+            is JavaVisibility.Protected,
+            is KotlinVisibility.Protected -> isAllowedInPackage(packageName, DokkaConfiguration.Visibility.PROTECTED)
+            is KotlinVisibility.Internal -> isAllowedInPackage(packageName, DokkaConfiguration.Visibility.INTERNAL)
+            is JavaVisibility.Default -> isAllowedInPackage(packageName, DokkaConfiguration.Visibility.PACKAGE)
+        }
+
+        private fun isAllowedInPackage(packageName: String?, visibility: DokkaConfiguration.Visibility): Boolean {
+            val packageOpts = packageName.takeIf { it != null }?.let { name ->
+                packageOptions.firstOrNull { Regex(it.matchingRegex).matches(name) }
+            }
+
+            val (documentedVisibilities, includeNonPublic) = when {
+                packageOpts != null -> packageOpts.documentedVisibilities to packageOpts.includeNonPublic
+                else -> globalOptions.documentedVisibilities to globalOptions.includeNonPublic
+            }
+
+            val isDocumentedVisibilitiesOverridden = documentedVisibilities !== DokkaDefaults.documentedVisibilities
+            return if (isDocumentedVisibilitiesOverridden) {
```

Comment:
Is `isDocumentedVisibilitiesOverridden` extra?

Reply:
What do you mean by extra? If you're proposing to inline it, then no, it's not extra - `isOverridden` in the name is supposed to explain the reason for reference equality check

Reply:
For example `documentedVisibilities.ifEmpty{ DokkaDefaults.documentedVisibilities }.contains(visibility) || includeNonPublic`?

Reply:
In this case we're imposing the default value on documented visibilities even if the user chose to override it with an `emptySet()`. The idea was that if the user has explicitly overridden it, we always prefer it and ignore deprecated `includeNonPublic`. If we try to mix them, it can lead to unexpected behaviour

I noticed an inconsistency with `visibility == PUBLIC` and pushed a fix, perhaps you'll like it more. Also left a comment with the explanation in case it wasn't obvious enough